### ISSUE TITLE
Changing IPFS Hashes

### ISSUE THREAD

**Issue by Scotty0448** [RavenProject/Ravencoin#336]
I was surprised some time ago when I first saw that the ability to change the IPFS hash was tied to the Reissuable setting. But a user could at least leave it set true if they thought they might want to change the IPFS hash sometime, so I never brought up any concern.

However, in the new 2.0.6 version I see that because it wouldn't make sense to reissue additional quantities for a Unique Asset, that setting is disabled and that means that any IPFS hash added is permanent. I think this could definitely be a problem.

Here are some situations I can think of right now where it would be desirable to change the IPFS hash, and not just for Unique Assets:

- The adminstrator makes a typo or other mistake in the metadata.
- Information like the contact name/address/email/phone in the metadata needs to be updated in the future.
- The asset is sold and the new owner wants to update some information like the For Sale setting or again any contact information.
- Links to external pages need to be updated if original destinations become unavailable.

It's understandable that some fields should never be changed once set, but some others certainly need to be editable in common and legitimate circumstances. I don't have a clearly best solution to suggest though, but here are some possibilities I can think of:

- Don't disable the Reissuable setting for Unique Assets and instead just leave the Quantity field disabled.

- Add an additional setting for each asset class so that being able to change the IPFS hash can be set independently.

- If one of those is implemented, also strongly encourage the addition of a Previous Hash field in the metadata to create a linked version history.

- Ideally create a protocol-level version history solution so that IPFS hashes always remain chageable but verson history is reliably kept.

**Comment by TronBlack** [comment]
This is a valid concern.  I'll raise this with the team on Monday.

**Comment by cfox** [comment]
tldr need to separate "reissue" from "change ipfs hash"

**Comment by stevenzeiler** [comment]
Yes, please we definitely need to be able to update the IPFS hash of our corporate security whenever the articles of incorporation filed with the state are updated.

+1 for `separate "reissue" from "change ipfs hash"`

**Comment by jeroz1** [comment]
I think with upcoming Open Index Protocol support, people should be able to substitute an IPFS hash for an OIP hash. Or will the OIP hash be an extra field?

**Comment by ravenlandpush** [comment]
unless there is some good mechanism to determine the difference between the two and no chance of collision then there may be a requirement for extra field?

**Comment by Scotty0448** [comment]
The new develop branch code is supporting OIP hashes in the same field as
IPFS hashes. They can be distinguished reliably by their differing lengths
(46 chars vs. 64 chars) and the code is only accepting those two lengths.
Using OIP alone doesn't solve Steven's issue though because OIP/FLO records
are immutable just like IPFS files.

But whereas the code requires that a 46 char hash must also be an IPFS
hash, a 64 char hash only has to be 64 hex chars and therefore can be
something other than an OIP hash. That would include any SHA-256 hash which
is commonly used throughout blockchain projects and other areas.

Having a way to change an asset's original hash regardless of whether the
asset is reissuable still seems like a good idea to me.  I particularly
like the idea that an asset's hash could point to the head of a linked list
of immutable data ("head" meaning the most recent item in the list).  That
way the list becomes an append-only version history of changes.  The items
in the list could be blockchain transactions themselves, each pointing back
to the previous transaction.  A validly-linked list could even be enforced
at the consensus level so that to change the asset's hash the new hash must
point to a transaction which contains the hash that is about to be
replaced.  So the spirit of making an asset no longer reissuable is
maintained while still allowing the hash to at least change if it always
points to an immutable version history.

Alternatively an asset's hash could point to the beginning of a linked
list, and therefore never need to change, but finding the most recent data
would require traversing the whole list every time and enforcing the
validity of the linked list become more problematic at the consensus level.

**Comment by jeroz1** [comment]
You can edit the records and the change history is stored on chain, much like a git. So that allows you to incrementally and transparently add information to tokens.

**Comment by TronBlack** [comment]
Having the promise be immutable and a fixed number of tokens has a value.  So I'd like to keep the IPFS hash immutable on non-reissuable tokens.

If we want to come up with a convention that an explorer could use to show a history of changes, then that could be done 2nd layer.  For example, sending a memo (IPFS hash) to RMetaDataUpdateXXXXXXgrzcxd could indicate a meta-data amendment (update) to a non-reissuable token.

Tron
________________________________

**Comment by jeroz1** [comment]
Ironically, I just mentioned to Scotty that having some kind of drop down showing the change history of an OIP record could be a nice new feature. Is this also an idea for in the wallet?

**Comment by ravenlandpush** [comment]
From a security aspect I quite like the idea of a hash not being changeable, like the assets json on the chain itself, although there is obviously many advantage to allowing ipfs field to be changeable in the chain seperately to the asset data. 

I think this requires some thought. Although it would be bad to change ipfs_hashes on assets which are not resissuable. It's a pretty bright idea to have a way to be able to have an asset which has re_issuable ipfs but not reissuable token quantity, etc.

 So a new flag is being suggested 'reissuable_ipfs_only'. IS this what your suggesting scotty? It's late here so hopefully this won't sound to much like a ramble. (its soooo going to sound like a ramble 😂 )

 Right now the two asset token details and ipfs_hash field are 'bulked together'. I think that is the crux of this discussion... should they be seperated, by a flag, or a different optional type?

 For some things such as deeds, contracts, wills, and so on. You want to make sure if its non-reissuable it stays that way right, but you might want to revise the contract for numerous legal or regulatory reasons too. I think point scotty makes is ipfs_hashes will need to be updated a lot more than asset token data, and if it wasn't so expensive to update both, maybe it could be offered to update one or another via new type of field 'changing_ipfs_contract'.  And naturally, this is additional RVN burnt from the use case that arises, that would have otherwise not been there, since many are very much dissuaded from updating ipfs_hashes in this fashion, from the expense of 500RVN for a main asset. And many others who do not want their asset to be reissuable for trust reasons, aren't able to update the chain of IPFS_hashes trust seperately. So there is a use case here...

 What would be nice, is to see an update_ipfs_hash_only, and it being represented for the proportion of price for a full-asset-details-change. That said, it could mean less RVN burnt for all the people that do reissue the whole asset, just for the purpose of updating the ipfs hash. Am I overthinking it?? I might be!!!

 But at the same time, deeds and wills, and testimony, they are chronological things, and their `witnessed` statements are and should be changeable seperately to the token themselves they are represented by , so why not represent them with ravencoin some way better? like a list of changes as scotty suggests? I can see in some use cases a need to seperate to the token data itself from the ipfs_hash. Some tokens need to stay the same, and their ipfs_contracts change, this seems not presently supportable by ravencoin, without OIP at least. It seems to me the best way to do that is to add an XOR for ipfs_hash, field like ipfs_hash_updateable_seperately , if OIP isnt enough. 

Here is an example Doug the lawyer, issues a client jon with their will, Jon requests his will to be updated, and Doug updates the unique asset ipfs_hash for 5 RVN using the controlling ! asset, since he used the 'reissuable_ipfs_cast_type'. This would result in extra RVN being burnt for UNIQUE updatable ipfs hashes, and I see a really important use case for what Scotty is suggesting after careful thought on the subject. It makes much more logical sense to do it like this, and keep that kind of data on chain, than perhaps relying on OIP to do it. That said someone may correct me that OIP sufficiently tracks this as desired for the use case. Certainly unique assets which are not reissuable, or use OIP is the way to go.

 There are many use cases where OIP or a reissuable ipfs_hash should not be used too. So I think its really important not to change existing functionality but 'support additional functionality'. And I see some cases where using ravencoin to track an seperately_updated_ipfs_hash stands to sensible reason. Also the cost of updating an ipfs_hash_only could/should be different, than say, reissuing an entire asset as aformentioned. This could introduce some more novel and potentially more economic use cases of ravencoin and the ipfs_hash. The question is would this introduce a lot of extra data/load to ravencoin network? Im not sure. It stands to reason to ask, why is the 5RVN unique hash not reissusable? When the extra 5 RVN burn could encourage more use of ravencoin ledger to track changes (As opposed to oip). I could be barking up the wrong tree here, maybe OIP does all this scotty.

But then again deeds/contracts/wills do need to be updated. The question should be less like will we allow ipfs_hash to be modified in old assets that are not reissuable and the answer more like 'is OIP enough for new assets?'' and tracking contracts?? 

I'm not 100% sure if OIP give you all temporal changes as ravencoin is capable, and maybe OIP already is better keep track of more of the changes, as it might be possible to do on ravencoin ledger. although OIP may itself keep some complex record of changes aswell which is equivalent (I have not looked too deeply into it). It makes sense here, what is being suggested is to keep a more complex track of changes to contracts, and allow ipfs hash to be potentially updated seperately to the rest of asset data, I feel that old assets should stay as they are,  but it is a nice idea to add this kind of ipfs_hash_only change, for those that dont want to use OIP, or might want to use the ravencoin ledger to properly track change instead of using OIP (and that is, if OIP can track these kind of changes) I'm not sure. I believe this is scotty's main point.

 We should not abandon that concept too quickly for a second layer solution, maybe. It would be nice say with the pop/push use case scotty talks about to provide perhaps a delta of difference between the last contract and the existing contract and have that on ravencoin ledger perhaps, especially if OIP not capable of this. Although perhaps this is also something more for a second layer. It could be useful utility/facility for establishing how much % the contract or file has changed thru the block timestamp, and other critical components when dealing with contracts. Or with the legal and regulatory aspects of it. I believe this might be very useful change if it is done carefully.

It seems to me that the original intent of the ipfs_hash was that if you say asset non resissuable, then the ipfs_hash should also not be changeable. 

I believe that old assets issued this way, should stay not changeable Then again nobody wants to 'waste' a good asset name, just because he or she can't update its ipfshash, or rushed and updated its ipfs_hash incorrectly first time. OIP would have helped there but it wasnt supported in the earlier versions, So of the 'crux' of this, it seems more a case of perhaps whether chain of contract changes happens natively on ravencoin, or via OIP, and to be honest I think I can see distinct advantages on both sides. (change of ipfs hashes data via ravencoin and changes of ipfs hash content via OIP). How would your 'second layer' metadata work tron with messages? Presumably the messages are stored permanently on the chain. For say people who had a no reissuable ipfs_hash and wanted to make it 'updateable? at the second layer. It's possible to do this on a second layer but everybody has to agree on how they do it. We need an RFC Standard for Ravencoin contracts. Of course it might be possible to add a new flag, set as NULL, for all assets, if everybody voted for it, that allows them to set ipfs_updatable_only and reissue_token_only, or reissue_both (as it stands presently).

I think we should be careful that one does not introduce a potentially dangerous precedent where information could be retracted or obscured from the blockchain - but this seems more of a utility than a hinderance to me. . At the same time we need to avoid a bunch of extra assets/fields being created/issued to deal with the same fact that some people need/want to re-issue their asests ipfs_hash, either because they have lost the file permanently or the data they posted was inaccurate.... So there are other reasons that give my support for this ... It makes sense to use OIP in future to prevent people being unable to change their IPFS HASH field... but for all existing owners. It might actually make sense to vote for a change if it is easily implementable to give people opportunity to set seperately ipfs_hash reissue to token reissue. And I can see many supporting that. It should be safe to add 'reissue ipfs_hash' at this time, but later on it would be bad, so now is the time to decide.

This could be a terrible suggestion, so if it is feel free to ignore me LOL!

Best,
Adam

**Comment by TronBlack** [comment]
Changing the IPFS for a security works now as long as you leave it reissuable.


Tron
________________________________

**Comment by Scotty0448** [comment]
Adam asks to verify if he's understanding my suggestion, but I haven't been
able to fully work something out to quite be considered a suggestion, and
probably shouldn't at this point anyways. There's still things I don't
fully understand. But I just see plenty of times when it will be desirable
to lock the quantity and units of an asset, to ensure no dilution can take
place, but not lock the data that's attached with the IPFS hash.

One idea I'm mentioning is that if the hash is left changeable then it
could be good to allow or ensure that a version history is kept in some
way. Or with Tron's idea, the hash could remain unchangeable but subsequent
special transactions could override it. It seems like that could work, but
a non-trivial change would need to be made to all explorers and some other
2nd or 3rd layer applications. Unfortunately anything that doesn't change
and handle those transactions properly will show an old hash and no
indication that it's no longer valid.

**Comment by ravenlandpush** [comment]
I think I might have overdone it on the length describing the problem. I think you put it quite well here 

> [at] times when it will be desirable to lock the quantity and units of an asset, to ensure no dilution can take place, but not lock the data that's attached with the IPFS hash.

My point , really coming from a 'security' standpoint, it would be good to identify if an asset ever has had it's ipfs_hash changed . Since for more transactions - it would be desirable to utilize the bitcoin. 

For instance if I want to verify a download that has been sha256summed (or GPG signed) from an ravencoin assets 'pointing' ipfs hash, I could.

`verifystring=$(raven-cli getassetdata RAVENCOINOS/BOOTSTRAP | grep ipfs_hash | awk {'print $2'} | sed 's/"//g' | xargs -i curl https://gateway.ravenland.org/ipfs/{})
downloadstring=$(sha256sum ravencoinOS-bootstrap.tar | awk '{print $1}');
if [ "$verifystring" == "$downloadstring" ]
then
// good
else
// not good
fi
`

This prompts the thought, maybe it would be good to have an rpc function that makes a lookup to see how many times the ipfs_hash changed. @blondfrogs is something like that possible? Having since spoken to @jeroz1 I'm made aware that OIP can, like github actually track quite well the changes of file states , so an OIP hash could theoretically do it all. It still prompts the question, from the security standpoint, maybe it would be good to have an RPC function to determine if the ipfs_hash has been'tampered', i.e. changed. This in itself seems like a good idea. As it stands I could reissue my GPG signature or sha256 sum, and all the people using the above code would then find themselves at the mercy of a tampered ipfs_hash, since the one referenced from the blockchain can be reissued.

There is a good case here , I think, to seperate ipfs_hash from the reissuing of an asset, for the reasons that you may want to reissue an asset, but for security purposes not want to reissue the ipfs_hash a 'fixed contract if you will' with non fixed issuance requirements. Maybe I am overthinking it a bit..

 (so not just the case of have a token that you dont want to reissue, and an ipfs hash you do, there is also the case above of where you have an ipfs_hash you dont want to reissue ever, for security purposes, but you may yet still want to reissue the asset token quantities etc. In the use case I give up, it would stand to purposeful reason to have an RPC function that is 'missing', and solves the aformentioned problem, either, thru the native getasset call for a field 'has_been_changed_before' added to the datatype. or  alternatively an RPC function, hasipfschangedbefore, that checks the previous blocks ipfs_hash and whether it was the same. I think in my usecase of using Ravencoin blockchain to verify say the bootstrap we have at ravenland.org for RavencoinOS, it could be improved in terms of its resistance to insecurity is to check two things, one of which can be achieved already, 1. check whether the asset is reissuable, then also fail the 'verification' comparison of filestring and signature string we retrieved from ravencoin rpc, if it is. What that doesn't encompass though is if it was reissuable for sometime in between for instance, so the only way to know for sure is to know the exact date, which then stands to reason why chronological timestamp against the ipfs_hash might be a useful function to provide in RPC, or at least some RPC function that can identify to my verifier script above, how many times the ipfs_hash changed in it's history, that way, you can guarantee the security quite easily then by adding a really simple way to validate from the native rpc commandset, for the most paranoid of us, when total Ipfs changes are under < 0 then trust the file. Otherwise I see situations where the utility use of the bitcoin ledger could be potentially lost, or at least made harder to achieve something as aformentioned. 

I do not know if I describe it well, but my opinion is that it's a fairly good idea to seperate ipfs_hash reissue from asset reissue, or , at least as a compromise, to provide some RPC function that can give a validation script for how many times that ipfs_hash has changed. It could be used for browsers to identify whether sites, or 'ravencoin assets' have been modified since they were 'last visited'. Perhaps this can be done natively already but my contribution to this. Cheers for suggestions Scotty I quite like them, and gave me a lot to think about 🤔 

Sorry for another essay 😂 

Best wishes,
push

**Comment by bedri** [comment]
I had read the whole messages and thought for some long time about this subject. Before I explain my ideas, I wanted to ask if we have any Asset Versioning concept (not related with Ravencoin Core versions but a versioning concept just about assets) in the codes.
If we don't I will write another long (maybe a separate article that may be a RIP on this) message about this subject but for short:
If we have/can have versioning in assets only, then every reissue will carry a new version and so that the IPFS hash or any change in the asset specifications can be versioned and thus kept on blockchain. Hence they can practically and on a consensus base be tracked securely. If this is thought or planned earlier, please let me know. If not, I will send a RIP request about this as soon as possible (please do feel free to send any RIP about this if you also light some bulbs without worrying about the idea licencing. For the record, all my ideas are under GPL last version license. Freedom first).

**Comment by MangoFarmAssets** [comment]
I think it is a good idea to consider (1) separating out assets reissuable (Y or N) and metadata reissuable (Y or N) in the chain data, and (2) adding an rpc call “reissue_metadata(...)” that would allow only the metadata hash to be changed - perhaps with a lower burn cost of say 5 RVN.  This would allow for immutability of one, the other or both - where desired.  Lots of conceivable use cases might need immutable quantity/units but still need to change the metadata over time.  Currently they have to keep the asset reissuable in order to preserve that ability.  And reissuing for that reason only (e.g, changing substantive documentation or even just changing the contact name in he JSON) might warrant a lower burn fee.  

Wallets and explorers can show the history of metadata changes if that is a concern.  

Food for thought.  I know we would use this functionality if it were adopted.

**Comment by TronBlack** [comment]
Has anyone looked at ipns?
https://docs.ipfs.io/guides/concepts/ipns/
https://medium.com/coinmonks/how-to-add-site-to-ipfs-and-ipns-f121b4cfc8ee

I seems like a way to change data and keep the anchor reference.

I don't yet understand the security on it.  Would it solve the same problem?

**Comment by ravenlandpush** [comment]
@TronBlack With IPNS is my understanding the same as IPFS but with mutability so it would result in the same problem I think.  The only way to ensure that the ipfs_hash is trustworthy is for the ipfs_hash to be immutable as well, and/or a history of ipfs changes on ravencoin ledger and there date may be sufficient with linked contract, IPNS is basically IPFS with mutability - that is my understanding anyway! 

Best wishes,
Adam

**Comment by MangoFarmAssets** [comment]
Tron - I think IPNS would be much like having an IPFS hash for a JSON that references an unsigned contract_url.  You could change the contents of the URL without changing the JSON (or the IPFS hash).  It works for that purpose but is less than ideal because the only immutable reference on chain would be the JSON hash, which proves nothing.  You could change it and no one would know.

**Comment by ravenlandpush** [comment]
Right doug! The IPFS Hash itself is immutable, the real contention is whether it is going to be by-design behaviour to allow an ipfs_hash to change on non re-issuable assets, and, if reissuable assets can have a non reissuable ipfs_hash.  Using OIP may not be a 'full' solution for the most 'secure' of transfers, unless it can be shown to be as reliable as bitcoin consensus.

It's important to note that IPFS itself is a blockchain, so a singular IPFS_HASH traditionally, write once, and (hopefully) 'read forever' it is immutable based on the sha hash of the file. 

The IPFS_HASH, if it is guaranteed to be 'write once, immutable', as a security feature works really nicely with Ravencoin and IPFS as it is, and it allows me to get a sha256sum, I know is definitely from the chain and a file that was placed there 'forever'.

 What is less apparent is how to get the number of changes made to that ipfs_hash, or rather that ravencoin assets pointing reference. If OIP as a single reference is added to ravencoin blockchain, and ravencoin blockchain itself cannot keep track of changes or provide a 'trusted guarantee', can OIP provide something equivalent?

 This is an important edge-case as to the signature file checker I give above it seems like a relevant subject to highlight, as there could be time in between reissuing the asset and reissuing/resetting the ipfs_hash. So there should be some assets that want an ipfs_hash reissue being seperate to the token issue is the point. Because otherwise reissues could happen, presently it should be possible with IPFS to track somehow in raw transaction data how many times the IPFS changed already, there is just no RPC function for it. Though - with OIP it is much less clear to me exactly how OIP guarantees the data it served didnt change for instance. Theoretically one might think or expect, there is a way to programatically refer to OIP so that if there is more than n changes, that it is not 'immutable' anymore.

Another point or question is can one use OIP and IPNS to do immutables? If so maybe my suggestions are redundant.


Best,
Adam

**Comment by lsji07** [comment]
Hi,

The immutable IPFS CID (content ID's) hash is only immutable if referencing a file.

However the way IPFS currently identifies IPFS nodes or peers are via the PID (peer ID). For IPNS as I understand it, the PID hash is used rather than the CID, and therefore pulls the files from that specific peer node when updated or "published".

Both IPFS CID's and and PID's use the same 46 letter Qm hash. 

So I believe this is a kind of irrelevant issue now. Both immutable and mutable data can be linked via Qm metadata field and this can be to the issuers requirements.

Not sure if this is helping or opening a can of worms.

**Comment by ravenlandpush** [comment]
So, this seems like a non issue if that is the case, for the most part. The point I'm making is do you rely on ravencoin blockchain to track ipfs hashes, or entirely rely on OIP. or compromise between the two. Ravencoin already can provide historic ipfs_hashes somehow thru raw transaction, I still think it would be nice for scottys suggestion to be implemented, of change of ipfs_hashes on blockchain, since changing hashes on blockchain, is possible, beyond merely changing filesystem on OIP. 

but you may have a point LSJ.

Best,
Adam

**Comment by bedri** [comment]
IPFS is a choice made by an exact need of a secure and immutable deposit of files. It's still a third party to get binded to Ravencoin Network so for instance (I don't think this will happen but still) if IPFS collapses, all asset specification database collapses. Now, adding another must-have-to-survive chain/structure like OIP will damage the security exponentially. Having OIP is good but it **must not be** a **have to**.

Having reissuable_ipfs_hash flag on an asset is a good solution as far as I see. For a larger solution I still think versioning assets is better. If we have a versioning convention for assets like version=0.0.0.0 where first digit is for asset reissue versions, second  digit is for ipfs_hash_reissue versions and the rest is for other kind of changes' versions it will carry more information about the history of assets with a less amount of data surveyed. A seperate leveldb database can be kept within Ravencoin network for to match the versions to the associated entities like ipfs hashes or reissued amounts as well as the timestamps. This last data database can be a pruned for light wallet cases.

IPNS is aimed to work for web sites hosted on IPFS so yes, with a single hash, even if the content of the website changes, with a DNS text record binded to IPNS hash, the site still continues to serve without changing the DNS text record.
I don't think this will serve our immutable single file content approach with a trackable history. But I may be missing something about IPNS approach of course.

**Comment by ravenlandpush** [comment]
Yeah. imagine if you had a ravencoin asset, that was reissuable, and you had historically added 10 different OIP' hashes. That's a valid point !

**Comment by MangoFarmAssets** [comment]
LSJ - good point.  You can definitely link to mutable data now without even using IPNS.  Just make a text file with http://ravenland.org in it, pin it to IPFS, and use the Qm hash as the on chain metadata.  That problem is solvable.  

The problem is nobody knows when you make a change to the Ravenland website this way (or, I think, the IPNS way).  

Having an IPFS hash on chain and reissuing with a new hash will show that the document was changed, but you can’t do that unless the whole asset is reissuable.  

I don’t know if the juice is worth the squeeze but allowing IPFS hashes to be changed on chain without reissuing the whole asset does seem like it would add some value.  Of course making the change is probably easier said than done!  

Cheers guys, 
Doug

**Comment by lsji07** [comment]
I would love to have some way of indicating which IPFS hashes are peer id's (changeable data) and which are content files (immutable). 

Unfortunately the only way I can think of is haven a check box when creating  for reissuing the tokens.

Wait we already have one of those in the reissue checkbox on asset token creation.

Only issue is that checkbox doesn't cover the immutable tokens Qm hashes as the IPFS hashes could be "reissuable" by the "trusted" issuer if created as PID type hashes. Granted only the original peer node (with their public and private keys) could reissue that token hash, but it is possible.

**Comment by ravenlandpush** [comment]
Peer ID's are hashes of the nodes publickey, so they should be formed already in a manner identifiable from IPFS_hash. As such you will not be able to retrieve peerid's thru a gateway, since they are used differently, and as such identfiable by the underlying ipfs code. Where, is anyones best guess.

  For example, the ravenland ravencoin IPFS private swarm we setup of these 4 nodes on AWS... the swarm connect uses the following ipfs/$peerid's. 
```
# connect to the Ravenland IPFS Private Ravencoin Swarm
ipfs swarm connect /ip4/91.134.145.4/tcp/5555/ipfs/QmWWA3RAuD9gN3fQBJFed9TZVC4J4BuQfcgfdqGgNNpw1x
# secondary swarm node 2
ipfs swarm connect /ip4/3.8.238.93/tcp/4001/ipfs/QmUZsTG8yMM8CnfQZhfGmaYRFTP37VwWnBuTuYbjhxSVAX
...
```

You'll find making a direct request to the peerid does not succeed at the gateway, since it not represent a file to serve thru gateway. This must be done by the signature of the hash in some way and the way it matches other peerid's connected in swarm.

 https://gateway.ravenland.org/ipfs/QmWWA3RAuD9gN3fQBJFed9TZVC4J4BuQfcgfdqGgNNpw1x 

These IPFS_hash are special in some way, and only are used to identify/discover the peerid's connection details based on the reverse hash of the string. So the way to tell the difference between IPFS hashes and peer id's already exist somewhere, I think. 

For me, although you raise some interesting idea LSJ, this is more a contention about the manner in which changes (timestamp) sensitive changes happen to ipfs_hash field in ravenland getasset details json, and when using OIP or a traditional ipfs_hash, whether ravencoin is at all involved in the tracking/versioning/history of those changes. ( and/Or whether it should be assumed a single OIP hash from inception of the asset is suitable for tracking such things. I certainly would prefer using bitcoin ledger and fixed ipfs_hashes with history on ravencoin chain for it but this is my opinion. OIP is a great way to add entire github histories. But it doesn't account for the full history of the asset, if it has been reissued etc) . So an OIP hash could have multiple issues. I dont see a way natively for OIP to keep track of it's previous OIP hashes to the blockchain, and I suppose we have to decide how common that will be, and whether the disadvantages of adding extra feature to the blockchain outweighs any potential advantage of tracking ipfs_hash changes in ravencoin (aswell as filesystem changes via OIP) et al.

It seems there are potential advantages/disadvantages for both. Some deep thought will likely be involved as to whether OIP is left to do this, or whether ravencoin raw transaction to check timestamp changes of ipfs_hash might be enough. Theoretically the ravencoin blockchain already keeping track of ipfs_hash changes, just not at native RPC functions. Correct me if im wrong, all previous ipfs_hash'es are stored already. It's simply a means of a call to berkley database , the question is how common this will actually be needed (if at all), certainly I can see advantages with file signature verification 'native' to the ravencoin blockchain. And certainly I see a couple of 'edge cases' where it would be useful for such a file signature ravencoin blockchain powered checker/verifier for ravencoin to have an extra check, so I can check the number of changes that it should have. This way it is possible on ravencoin ledger alone, to determine whether the ipfs_hash retrieved is the one we expect. Which for some security uses could be important. Again if we know which OIP hash, and the first time of the data we are looking to validate already, one might be able to rewrite the function to use OIP (on 2nd layer). So its really whether its going to rely on second layer, and/or somewhere in between and perhaps provide an rpc function to give ipfs_hash_times_changed. a Useful reference point for fixed/static reference IPFS_HASH since you know with certainty it won't be changed, without relying on OIP. But I think my point is it can probably be achieved with both. What OIP wouldnt be aware of, likely, is changes before the OIP hash was added, and that could also be significant!

Cheers &
Best wishes
Adam

**Comment by TronBlack** [comment]
Here's the reasoning behind combining the ability to change the IPFS hash, and issue more tokens together and only have one binary flag for reissuable/immutable.

A primary goal of Ravencoin is simplicity -- Keep It Simple Stupid (KISS).  Having a single flag (true/false) (reissuable/immutable) that indicates that what you see is what you get is simple.

Current:
TRUE - You can increase qty, change units to more divisible, mutable metadata
FALSE - Locked qty, Locked Units, Locked metadata

Suggested:
TRUE/TRUE - Increase qty, change units to more divisible, mutable metadata  (same as TRUE above)
TRUE/FALSE - Increase qty, change units to more divisible, locked metadata  (NEW)
FALSE/TRUE - Locked qty, Locked Units, mutable metadata  (NEW)
FALSE/FALSE - Locked qty, Locked Units, Locked metadata  (same as FALSE above)

Keeping in mind the goal of simplicity, let's look at the two use-cases that we add and see if they're worth complicating the protocol.  Also, keep in mind that the owner token holder (Issuer) is the only one that can make these changes.

For the TRUE/FALSE case, the description of what the token represents (metadata) is fixed, but the issuer can create millions or billions more of the token.  In this case, the token holders are trusting the issuer not to oversubscribe, water-down, or dilute existing holders.  If you're already trusting the issuer for this, how important is it that they can't change the metadata - making it the similar to reissuable now.

For the FALSE/TRUE case, the qty is locked, and units are locked, but the metadata describing the token can be changed by the issuer.  So maybe you're guaranteed that there are only 1 million tokens that represented a 1/1,000,000 share of a rare Rembrandt painting as described by the metadata.   Now the metadata changes to 1/1,000,000 of a candy corn (the ones nobody likes) because the metadata can be changed.  What good is locking the qty and units if what the token represents can be changed at will by the issuer?  If the issuer is trusted not to change what the token represents, then leave the token reissuable. 

Note that in every case, whether we switch it to two flags because we find important use-cases or stays as a single flag, any changes to the qty, units, or metadata will be detectable changes and all changes will be immutably recorded.

I personally prefer a simple single flag so that once the token has been marked as non-reissuable you get what you see and know it can't change - like Bitcoin.  If reissuable, you know things can change (qty, units, metadata).

If there are compelling reasons to switch to two flags with good real-world use cases, my opinion could be swayed.

**Comment by MangoFarmAssets** [comment]
Hi Tron.  Those are all good points.  

On the flip side, you could view it more as adding flexibility than complexity.  There are lots of assets you could conceive of where the issuer (and purchasers) would want the quantity to be immutable but allow flexibility with the metadata.  

Just for example, take a hard asset like a bar of gold.  I want to fractionalize it immutably into 100 one percent shares and sell each share.  My metadata has a contract that says the gold will be stored in warehouse x, an contains an obligation to maintain insurance.  I attach an insurance policy for 2019-2020.  The only way I can add the insurance policy for 2020-21 next year is by keeping the asset reissuable, and the quantity mutable.  I bet we could spitball hundreds of examples like this.

I view the IPFS hash almost like a repository for any documentation, or even any information that can be stored in a DB.  That information could - but doesn’t necessarily - tie to the quantity of the asset itself.  As one very practical example, currently you can’t change your contact email address in the metadata if an asset isn’t reissuable.

**Comment by TronBlack** [comment]
Using your example...

The same trust/ethical-behavior/social-norms/regulations/threat-of-prison that prevents the issuer from changing the metadata from gold to candy corn (instead of updating to 2020-21 insurance), should prevent them from changing the issued qty from 100 to 1000.  Either change is transparent on-chain.  So, in this use case, the token should remain reissuable - so the issuer can update the contract.

Is there a danger of the issuer changing the qty, or a benefit of locking the qty, if the issuer could also just change the metadata to say that each of the 100 tokens now represents 0.1% of the bar of gold vs. 1% of the bar of gold.  It has exactly the same net effect.

I guess my thinking is that qty issued and token divisibility really are just metadata, and only separated out so that the blockchain can handle the accounting.  If the issuer can change the nature of what the token represents, how is it more valuable or secure that they can't change the qty of X (where the nature of X _can_ be changed)?

Why I'm arguing this case with a seasoned litigator is beyond me.

**Comment by MangoFarmAssets** [comment]
Ha!  Probably the same reason I am on GitHub talking to a principal software developer.  Fair point.  You are 100% right that the very nature of the thing can be changed with any change in metadata.  So this isn’t really a “solution” (and I’m not sure it is a problem either).

**Comment by Scotty0448** [comment]
It's very good to read some more extensive thoughts about the issue.  I can
tell that what's been keeping Tron and myself at least from seeing things
more similarly is that Tron has been focusing more on the metadata fields
that describe what the asset is.  But I've been focusing more on the
secondary information about the asset that could commonly change, such as
contact information, physical location, website URL, or something that
needs to regularly change like Mango Farm's insurance policy example.

So it's more clear to me now how the issue exists because there is only
just one hash field and yet it could be desirable to associate both mutable
and immutable data with an asset.

Perhaps the best solution is to leave things as they are and just assume
that making an asset non-reissuable should be fairly rare and only when the
metadata can be limited to only fields that should never change.  Other
cases would just trust the asset owner to not materially change what the
asset is even though additional information might be changed.  As has been
pointed out many times, trusting the issuer is most always necessary
anyways and a blockchain implementation isn't going to affect that.

The only remaining idea I'll throw out is that possibly having two hashes,
one for mutable data and one for immutable data, could solve the issue
while also keeping things fairly simple and understandable.  Neither hash
would even be tied to the reissuable flag, which would only affect issuing
additional quantity of an asset or its units.  This may be too much to ask
for a first layer change, although I believe it at least wouldn't be very
complex. I realize it also adds additional size to the blockchain in cases
where an asset is assigned both hashes.  So maybe it's a non-starter but I
thought I'd throw it out just in case the issue remains and options are
being looked for.
